HEAD should always be supported by mapping to GET when supported

**Expected behavior**
Views should always support the `HEAD` method by mapping it to the `GET` method, *when supported*.

**Actual behavior**
If the function-based view does not have `HEAD` in its methods or the class-based view does not define `.head()`, the `HEAD` method is interpreted as unsupported and a `405` error is returned.

**To Reproduce**

```python
import bocadillo

api = bocadillo.API()

@api.route('/', methods=['get'])
async def index(req, res):
    pass

response = api.client.head('/')
assert response.status_code == 200
```

**Screenshots/Traceback**
```
>       assert response.status_code == 200
E       assert 405 == 200
E        +  where 405 = <Response [405]>.status_code
```

**Specs**
- OS: macOS 10.11
- Python Version: 4.7
- Bocadillo Version: master

**Additional context**

`HEAD` requests are used to obtain meta info about a resource without transferring the whole body, and are aimed at being identical to `GET` but without a body (source: [w3.org](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)).

Web servers automatically strip down the body of `HEAD` responses to comply with the HTTP specs. So handling a `HEAD` request as  `GET` (i.e. returning a body) should be fine. Plus, some systems rely on `HEAD` to check for existence of the resource (e.g. web crawlers), so we should always be supporting `HEAD` even if not specified (source: [a note in the Django docs](https://docs.djangoproject.com/en/2.1/topics/http/decorators/#django.views.decorators.http.require_safe)).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -27,6 +27,7 @@ Bocadillo adheres to [Semantic Versioning](https://semver.org).
 ### Fixed
 
 - Errors returned by custom error handlers could have 200 status in case the handler did not set any status code. It now defaults to 500.
+- If `GET` is supported, `HEAD` will automatically be implemented.
 
 ## [v0.6.1] - 2018-12-04
 

--- a/bocadillo/error_handlers.py
+++ b/bocadillo/error_handlers.py
@@ -1,4 +1,5 @@
 """Built-in error handlers."""
+import traceback
 from functools import wraps
 from typing import Callable
 
@@ -19,6 +20,8 @@ async def _res_for_exc(req: Request, exc: Exception, **kwargs):
     if isinstance(exc, HTTPError):
         res = Response(req, **kwargs)
         _handle_http_error(req, res, exc)
+        if exc.status_code == 500:
+            traceback.print_exc()
     else:
         res = await _res_for_exc(req, HTTPError(500), **kwargs)
     return res

--- a/bocadillo/routing/router.py
+++ b/bocadillo/routing/router.py
@@ -19,6 +19,15 @@ class RouteMatch(NamedTuple):
     params: dict
 
 
+def _ensure_head_if_get(view: View, methods: List[str]) -> None:
+    if inspect.isclass(view):
+        if hasattr(view, 'get') and not hasattr(view, 'head'):
+            view.head = view.get
+    else:
+        if 'GET' in methods and 'HEAD' not in methods:
+            methods.append('HEAD')
+
+
 class Router:
     """A collection of routes."""
 
@@ -48,6 +57,8 @@ class Router:
 
         if namespace is not None:
             name = namespace + ':' + name
+
+        _ensure_head_if_get(view, methods)
 
         if inspect.isclass(view):
             view = view()

--- a/bocadillo/view.py
+++ b/bocadillo/view.py
@@ -26,6 +26,12 @@ class ClassBasedView:
     def delete(self, req: Request, res: Response, **kwargs):
         raise NotImplementedError
 
+    def head(self, req: Request, res: Response, **kwargs):
+        raise NotImplementedError
+
+    def options(self, req: Request, res: Response, **kwargs):
+        raise NotImplementedError
+
 
 # Types
 AsyncView = Callable[[Request, Response, dict], Coroutine]

--- a/docs/topics/request-handling/routes-url-design.md
+++ b/docs/topics/request-handling/routes-url-design.md
@@ -140,17 +140,26 @@ you can use the `methods` argument to `@api.route()` to specify the set of
 HTTP methods being exposed:
 
 ```python
-@api.route('/', methods=['get'])
-async def index(req, res):
-    res.text = "Come GET me, bro"
+@api.route('/posts/{pk}', methods=['delete'])
+async def delete_blog_post(req, res, pk):
+    res.status_code = 204
 ```
 
-When a non-allowed HTTP method is used by a client, a `405 Not Allowed` error response is automatically returned. Callbacks such as [hooks] and [middleware] callbacks will not be called either.
+Methods passed to `@api.route()` are case-insensitive.
 
-::: tip
+::: tip NOTE
 The `methods` argument is ignored on [class-based views](../features/views.md#class-based-views). You should instead decide which methods are implemented on the class to control
 the exposition of HTTP methods.
 :::
+
+### How are unsupported methods handled?
+
+When a non-allowed HTTP method is used by a client, a `405 Not Allowed` error response is automatically returned. [Hooks] callbacks will not be called either (but request [middleware] will).
+
+::: tip
+Bocadillo implements the `HEAD` method automatically if your route supports `GET`. It is safe and systems such as URL checkers may use it to access your application without transferring the full request body.
+:::
+
 
 [Request]: requests.md
 [Response]: responses.md

--- a/tests/test_route_methods.py
+++ b/tests/test_route_methods.py
@@ -38,3 +38,20 @@ def test_allowed_method_must_be_valid_http_method(builder: RouteBuilder):
         builder.function_based('/', methods=['foo'])
 
     builder.class_based('/', methods=['bar'])
+
+
+def test_if_get_implemented_by_class_based_view_then_head_mapped(api: API):
+    @api.route('/')
+    class Index:
+        async def get(self, req, res):
+            pass
+
+    assert api.client.head('/').status_code == 200
+
+
+def test_if_get_in_function_view_methods_then_head_mapped(api: API):
+    @api.route('/', methods=['get'])
+    async def index(req, res):
+        pass
+
+    assert api.client.head('/').status_code == 200

